apps/bestofjs-nextjs/src/components/search-palette/search-palette-state.ts

```diff
@@ -0,0 +1,160 @@
+"use client";
+
+import * as React from "react";
+import { useRouter } from "next/navigation";
+import invariant from "tiny-invariant";
+import useDebouncedCallback from "use-debounce/lib/useDebouncedCallback";
+
+import { stateToQueryString } from "@/components/project-list/navigation-state";
+import { useSearchState } from "@/components/project-list/search-state";
+
+import { ClientSearch } from "./search-container";
+
+export type SelectedItem =
+  | {
+      type: "project";
+      value: BestOfJS.SearchIndexProject;
+    }
+  | {
+      type: "tag";
+      value: BestOfJS.Tag[];
+    }
+  | {
+      type: "text";
+    };
+
+const DEBOUNCE_DELAY = 300;
+
+export function useSearchPaletteTags() {
+  const searchState = useSearchState();
+  const { lookupTag } = ClientSearch.useContainer();
+  const [currentTagCodes, setCurrentTagCodes] = React.useState<string[]>(
+    searchState.tags
+  );
+
+  // The search palette is mounted only once, we need to sync the tags when the URL changes
+  React.useEffect(() => {
+    setCurrentTagCodes(searchState.tags);
+  }, [JSON.stringify(searchState.tags)]); // eslint-disable-line react-hooks/exhaustive-deps
+
+  const removeTag = (tagCode: string) =>
+    setCurrentTagCodes((state) => state.filter((tag) => tag !== tagCode));
+  const resetCurrentTags = () => setCurrentTagCodes(searchState.tags);
+
+  const currentTags = currentTagCodes
+    .map(lookupTag)
+    .filter(Boolean) as BestOfJS.Tag[];
+
+  return {
+    currentTags,
+    currentTagCodes,
+    removeTag,
+    resetCurrentTags,
+    searchState,
+  };
+}
+
+export function useSearchPaletteState() {
+  const router = useRouter();
+  const [isPending, startTransition] = React.useTransition();
+  const { lookupProject, lookupTag } = ClientSearch.useContainer();
+
+  const [selectedItem, setSelectedItem] = React.useState<
+    SelectedItem | undefined
+  >();
+
+  const [open, setOpen] = React.useState(false);
+
+  const [searchQuery, setSearchQuery] = React.useState<string>("");
+  const {
+    currentTags,
+    currentTagCodes,
+    removeTag,
+    resetCurrentTags,
+    searchState,
+  } = useSearchPaletteTags();
+
+  const onOpenChange = (value: boolean) => {
+    if (!value) {
+      resetCurrentTags();
+      setSearchQuery("");
+    }
+    setOpen(value);
+  };
+
+  React.useEffect(() => {
```

Comment:
We might want to add a comment summarizing this useEffect.

Or refactor the logic outside into a function like `_enableSearchPaletteShortcutListener`

Reply:
Good point, each non-trivial `useEffect` should be extracted in a custom hook, `useKeyboardShortcut()` for example.